### ISSUE TITLE
[i2c] Inefficient use of memory when saving devices

### ISSUE THREAD

**Issue by andreeaflorescu** [rust-vmm/vhost-device#27]
Devices need to be unique. To achieve this, they're defined as `device_map: [u32; MAX_I2C_VDEV]`. This is then initialized as follows:

```rust
let mut device_map: [u32; MAX_I2C_VDEV] = [I2C_INVALID_ADAPTER; MAX_I2C_VDEV];
```

The value of `MAX_I2C_VDEV` is `1 << 7` which makes for an inefficient memory usage as this vector is sparse (only a few elements are valid). A better structure for defining this is a `HashMap` if we need to remember indexes, or a `HashSet` otherwise.

**Comment by vireshk** [comment]
Pasting below my last message i sent on this during initial reviews.. I haven't tried it again..

**Comment by vireshk** [comment]
Hmm, I tried to do this but wasn't able to make it work. I tried this way (based on new patchset I just pushed, stuff got renamed):

pub struct I2cMap<A: I2cAdapterTrait> {
adapters: HashMap<usize, &A>,
}

I had to use an reference here as I need to use same Adapter structure for multiple client devices, since they will all be accessed via the same Adapter (which has a File descriptor as well).

It then got into timeslice issues, which ones fixed got me to the fact that I am adding a local I2cAdapter variable as reference to the hashmap and got this error "returns a value referencing data owned by the current function".

I need to make sure somehow that the I2cAdapter allocated in this function (that returns the hashmap) live long enough. One way is adding them to another vector inside I2cMap struct, but that will take us to where we started from :(

All in all, I couldn't figure out how to get references stored in a hashmap to work.

@juntian I hope I have fixed all the issues you pointed out in the previous version, thanks a lot for your reviews.
